### PR TITLE
Batch save blocks for initial sync. 80% faster BPS

### PR DESCRIPTION
**Pull request by terencechain**
This PR saves blocks in batch during initial sync. It holds the blocks in cache and batch saves the cached blocks into DB when there's a new finalized check point or the blocks count exceeds 64 (whichever comes first)

This is gated behind feature flag `--init-sync-batch-save-blocks`, I'm seeing on average **80-100%** faster blocks per second versus current initial syncing. 

This works and tested with the following flags:
```
--enable-ssz-cache 
--enable-state-field-trie 
--enable-byte-mempool
--enable-initial-sync-queue
```

I dont think it's compatible with new state mgmt flag yet

### CODE REVIEW

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 1]
```diff
@@ -170,10 +172,23 @@ func (s *Service) generateState(ctx context.Context, startRoot [32]byte, endRoot
...
 		return nil, errors.New("finalized state does not exist in db")
 	}
-	endBlock, err := s.beaconDB.Block(ctx, endRoot)
-	if err != nil {
-		return nil, err
+
+	var endBlock *ethpb.SignedBeaconBlock
+	if featureconfig.Get().InitSyncBatchSaveBlocks && s.hasInitSyncBlock(endRoot) {
+		endBlock = s.getInitSyncBlock(endRoot)
+		if featureconfig.Get().InitSyncBatchSaveBlocks {
```
Comment (0xKiwi) [inline beacon-chain/blockchain/init_sync_process_block.go]: Why is this condition here? Wouldn't it already be true?

Comment (nisdas) [inline beacon-chain/blockchain/init_sync_process_block.go]: why check again ? this will always be true

File: beacon-chain/db/kv/blocks.go [hunk 2]
```diff
@@ -227,16 +226,16 @@ func (k *Store) SaveBlocks(ctx context.Context, blocks []*ethpb.SignedBeaconBloc
...
 
 	return k.db.Update(func(tx *bolt.Tx) error {
+		bkt := tx.Bucket(blocksBucket)
 		for _, block := range blocks {
 			if err := k.setBlockSlotBitField(ctx, tx, block.Block.Slot); err != nil {
 				return err
 			}
-
-			blockRoot, err := ssz.HashTreeRoot(block.Block)
+			blockRoot, err := stateutil.BlockRoot(block.Block)
```
Comment (0xKiwi) [inline beacon-chain/db/kv/blocks.go]: Nice

File: beacon-chain/sync/initial-sync-old/round_robin.go [hunk 3]
```diff
@@ -193,7 +203,7 @@ func (s *Service) roundRobinSync(genesis time.Time) error {
 			1,                    // step
 			blockBatchSize,       // count
 			peers,                // peers
-			0,                    // remainder
+			0,                    // reminder
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync-old/round_robin.go]: Reminder? What does that mean in this context?

File: beacon-chain/sync/initial-sync-old/round_robin.go [hunk 4]
```diff
@@ -209,10 +219,12 @@ func (s *Service) roundRobinSync(genesis time.Time) error {
 
 		for _, blk := range blocks {
 			s.logSyncStatus(genesis, blk.Block, peers, counter)
-			if !s.db.HasBlock(ctx, bytesutil.ToBytes32(blk.Block.ParentRoot)) {
-				log.Debugf("Beacon node doesn't have a block in db with root %#x", blk.Block.ParentRoot)
+			parentRoot := bytesutil.ToBytes32(blk.Block.ParentRoot)
+			if !s.db.HasBlock(ctx, parentRoot) && !s.chain.HasInitSyncBlock(parentRoot) {
+				log.Debugf("Beacon node doesn't have a block in db or cache with root %#x", parentRoot)
```
Comment (0xKiwi) [inline beacon-chain/sync/initial-sync-old/round_robin.go]: Should this be changed to `WithField()` for better log tracking?

File: beacon-chain/blockchain/init_sync_process_block.go [hunk 5]
```diff
@@ -189,3 +204,48 @@ func (s *Service) generateState(ctx context.Context, startRoot [32]byte, endRoot
 	}
 	return postState, nil
 }
+
+// This saves a beacon block to the initial sync blocks cache.
+func (s *Service) saveInitSyncBlock(r [32]byte, b *ethpb.SignedBeaconBlock) {
+	s.initSyncBlocksLock.Lock()
+	defer s.initSyncBlocksLock.Unlock()
```
Comment (nisdas) [inline beacon-chain/blockchain/init_sync_process_block.go]: do we need a lock for this ? initial sync is single-threaded , so there isnt a risk of concurrent writes

File: beacon-chain/sync/initial-sync-old/round_robin.go (hunk 4 above)
Comment (0xKiwi) [reply beacon-chain/sync/initial-sync-old/round_robin.go]: Also, not feature flag needed here?

File: beacon-chain/blockchain/process_block.go [hunk 6]
```diff
@@ -247,6 +252,14 @@ func (s *Service) onBlockInitialSyncStateTransition(ctx context.Context, signed
 		}
 	}
 
+	// Rate limit how many blocks (2 epochs worth of blocks) a node keeps in the memory.
+	if len(s.getInitSyncBlocks()) > 2*int(params.BeaconConfig().SlotsPerEpoch) {
```
Comment (nisdas) [inline beacon-chain/blockchain/process_block.go]: can we make this value a configurable constant at the start of the file

File: beacon-chain/blockchain/init_sync_process_block.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/blockchain/init_sync_process_block.go]: Oops, thanks!

File: beacon-chain/sync/initial-sync-old/round_robin.go (hunk 3 above)
Comment (terencechain) [reply beacon-chain/sync/initial-sync-old/round_robin.go]: Sorry it should have been remainder

File: beacon-chain/sync/initial-sync-old/round_robin.go (hunk 4 above)
Comment (terencechain) [reply beacon-chain/sync/initial-sync-old/round_robin.go]: Updated to WithField(), feature flag not needed in this csae

File: beacon-chain/blockchain/init_sync_process_block.go (hunk 1 above)
Comment (terencechain) [reply beacon-chain/blockchain/init_sync_process_block.go]: fixed!

File: beacon-chain/blockchain/init_sync_process_block.go (hunk 5 above)
Comment (terencechain) [reply beacon-chain/blockchain/init_sync_process_block.go]: I tried without the lock, the performance is neglectable. Just figure be safe than sorry later if initial sync moves to multi threaded

File: beacon-chain/blockchain/process_block.go (hunk 6 above)
Comment (terencechain) [reply beacon-chain/blockchain/process_block.go]: yep, done!

**Pull request merged**